### PR TITLE
feat: add advanced example of contract state migration

### PR DESCRIPTION
**Pull request by denbite**
This example was built inspired by [DevHub Contract](https://github.com/NEAR-DevHub/neardevhub-contract/blob/main/src/migrations.rs)
I believe this elegant and scalable approach is the way to go for contracts that often roll out breaking changes to their contract's code on mainnet

### CODE REVIEW

**Comment by frol** [comment]
@denbite I love it! Thank you for extracting the best practice and documenting it clearly

@race-of-sloths score 13

**Comment by race-of-sloths** [comment]
@denbite Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Do you want to apply for monthly streak? Get 8+ score for a single PR this month and receive boost for race-of-sloths!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://race-of-sloths.com/images/bot-banner-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://race-of-sloths.com/images/bot-banner-light.svg">
    Shows inviting banner with latest news.
</picture>](https://race-of-sloths.com/latest-news)

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/denbite?type=bot&pr=near-examples/update-migrate-rust/2&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/denbite?type=bot&pr=near-examples/update-migrate-rust/2&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/denbite)








<details><summary>Current status: <i>executed</i></summary>

| Reviewer | Score |
|--------|--------|
| @akorchyn  | 13 |


Your contribution is much appreciated with a final score of 13!
You have received 150 (130 base + 20 monthly bonus) Sloth points for this contribution

@akorchyn received 25 Sloth Points for reviewing and scoring this pull request.

Congratulations @denbite! Your PR was highly scored and you completed another monthly streak! To keep your monthly streak make another pull request next month and get 8+ score for it
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/denbite)

For maintainers:
- Score pull requests that participate in the Race of Sloths and receive a reward
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: advanced-multi-version-updates/v3/README.md
Comment (akorchyn) [inline advanced-multi-version-updates/v3/README.md:47]: I think this won't work. You have the same StorageKey in use. So they will be resolved in the  same storage key and you end up with both data (as you are pushing new in)

**Review by akorchyn — APPROVED** [review]

File: advanced-multi-version-updates/v3/README.md
Comment (denbite) [reply advanced-multi-version-updates/v3/README.md:47]: I checked all migrations on [testnet](https://testnet.nearblocks.io/address/update-migrate-rust-advanced-multiversion-updates.testnet) - no issues found

New instance of `Vector` is created without knowledge of past data, and its length is set to zero in constructor. When a first message is pushed, it just overwrites the value at index zero and bumps the length.

Comment (denbite) [reply advanced-multi-version-updates/v3/README.md:47]: also, to prove it works, the contract I'm referencing to, returns a list of messages as follows (after V3 version was rolled out)
`near contract call-function as-read-only update-migrate-rust-advanced-multiversion-updates.testnet get_messages json-args {} network-config testnet now`

```
Result:
[
  {
    "payment": "100000000000000000000000",
    "premium": true,
    "sender": "denbite.testnet",
    "text": "a message"
  }
]
```

Comment (akorchyn) [reply advanced-multi-version-updates/v3/README.md:47]: Try to add an extra element in step 2 and you will face the error.
It looks like it's working but it is not. 
You have a Vector with 3 elements: 
```
|Element A|Element B|Element C|
```
                                                       
When you create a Vector::new(SameStorageKey)

You use the same memory but its index point is zero.
```
|Element A|Element B|Element C|
/\
|
```

So, you have two vectors. Both are pointing to the same place. When you iterate through the first vector and push into the second, you replace Elements from the left.
It would probably be fine if your new element size is less or equal to the previous one. But it would fail otherwise.
But even in this case, you might face the issue of still paying for space you don't use.
```
|Updated ELEMENT A IS BIGGER THAN PREVIOUS| ent B|Element C|
/\
|
```
After this push, you will have a screwed next element. But it will still try to read the data and might even succeed. In the end, you will end up either with a failing migration or screwed data that is way worse :(

Second iteration would look like that:
```
|Updated ELEMENT A |IS BIGGER THAN PREVIOUS ent B|Element C|
                                          /\
                                          | oops we have some mess of old second and part of new first
```

Comment (akorchyn) [reply advanced-multi-version-updates/v3/README.md:47]: Here is my V2 before migrating to V3:
```
[
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 4"
  },
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 3"
  },
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 2"
  },
  {
    "premium": false,
    "sender": "unnatural-error.testnet",
    "text": "pretty huge text will see what happens 1"
  }
]
```

Deployed V3:
```
handler error: [Function call returned an error: wasm execution failed with error: HostError(GuestPanic { panic_msg: "Cannot deserialize the contract state." })]
```

Migrated to V3:
```
   2: handler error: [Function call returned an error: wasm execution failed with error: HostError(GuestPanic { panic_msg: "Cannot deserialize element" })]
```

Comment (akorchyn) [reply advanced-multi-version-updates/v3/README.md:47]: Also, I'm not 100% confident about how StorageKey is generated, but I feel like it's not enough to replace a name, you need a new entry to be 100% safe. (e.g. change name to __RESERVED for example, and add new storage key. After you have cleared the entry, you might use this reserved key for other structs). Otherwise, you will overwrite the payments vector and might mess in a loop of that data.

@frol is the stated above correct?

**Review by akorchyn — CHANGES_REQUESTED** [review]

File: advanced-multi-version-updates/v3/README.md
Comment (akorchyn) [reply advanced-multi-version-updates/v3/README.md:47]: Also, I'm not 100% confident about how StorageKey is generated, but I feel like it's not enough to replace a name, you need a new entry to be 100% safe. (e.g. change the Payments key to __RESERVED for example, and add a new storage key. After you have cleared the entry, you might use this reserved key for other structs). Otherwise, you will overwrite the payments vector and might mess that data.

@frol is the stated above correct?

**Comment by race-of-sloths** [comment]
### ❌ PR inactive
This PR has been inactive for two weeks. It's now marked as stale. 
>[!TIP]
If you think it's a mistake, you can restart the bot with `include` command to proceed 🔄

File: advanced-multi-version-updates/v3/README.md
Comment (akorchyn) [reply advanced-multi-version-updates/v3/README.md:47]: Given the fact that the contract uses `state_write` as well V3 structs are cached and write on drop only. It works just fine. I'm not sure where my error is coming from as I can no longer reproduce it.

**Comment by denbite** [comment]
@race-of-sloths include

**Comment by akorchyn** [comment]
@race-of-sloths score 13

**Pull request merged**
